File: 1036-rotting-oranges/rotting-oranges.py
from collections import deque
from typing import List
dirs = [(1,0),(-1,0),(0,1),(0,-1)]
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        def isValid(r,c): return 0<=r<ROWS and 0<=c<COLS
        minTime = 0
        seen = set()
        q = deque([])
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == 2:
                    q.append((row,col,0))
                    seen.add((row,col))
        while q:
            currX,currY,currD = q.popleft()
            for x,y in dirs:
                newX,newY = currX+x,currY+y
                if not isValid(newX,newY): continue
                if grid[newX][newY] == 0: continue
                if (newX,newY) in seen: continue
                seen.add((newX,newY))
                minTime = max(minTime,currD+1)
                grid[newX][newY] = 2
                q.append((newX, newY, currD + 1))


        for num in grid:
            if all(True if n == 2 or n == 0 else False for n in num):continue
            else: return -1
        return minTime

File: 1036-rotting-oranges/rotting-oranges.cpp
#include <bits/stdc++.h>
using namespace std;

template <typename T>
void printUnorderedSet(const unordered_set<T>& s) {
    bool first = true;
    for (const auto& elem : s) {
        if (!first) cout << " ";
        cout << elem;
        first = false;
    }
    cout << '\n';
}

template <typename T>
void printVector(const vector<T>& vec) {
    for (size_t i = 0; i < vec.size(); ++i) {
        cout << vec[i] << (i + 1 < vec.size() ? " " : "");
    }
    cout << '\n';
}

class Solution {
public:
    vector<vector<int>> dir = { {1,0}, {-1,0}, {0,1}, {0,-1}};
    int orangesRotting(vector<vector<int>>& grid) {
        int minT = 0;
        queue<tuple<int, int, int>> q; // (row, col, time)

        int rows = grid.size();
        int cols = grid[0].size();
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                if (grid[r][c] == 2) { q.push({r, c, 0}); }
            }
        }
        while (!q.empty()){
            tuple<int, int, int> top = q.front();q.pop();
            int row = get<0>(top);
            int col = get<1>(top);
            grid[row][col] = 2;
            int currT = get<2>(top);
            minT = max(minT,currT);
            for (int i = 0; i<4; i++){
                int x = row + dir[i][0];
                int y = col + dir[i][1];
                if (x>=0 && x<rows && y>=0 && y<cols && grid[x][y] == 1){
                    q.push({x,y,currT+1});
                    grid[x][y] = 2;
                }
            }

        }
        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                if (grid[r][c] == 1) { return -1; }
            }
        } 
        return minT;
    }
};